File: Javascript/1-Two_Sum.js
// ---------- JAVASCRIPT ----------
var twoSum = function (nums, target) {
  // O(n) time | O(n) space

  // Hashmap to store numbers and their indexes
  const storage = {}; // O(n) space

  for (let i = 0; i < nums.length; i++) {
    // O(n) time
    // complement -> Number needed to add to current
    // element to reach the target
    const complement = target - nums[i];
    // If the complement is in the storage we can reach the
    // target -> RETURN indexes
    if (complement in storage) return [storage[complement], i];

    // Store current element with his index
    storage[nums[i]] = i;
  }
};

File: Python/1-Two_Sum.py
# ---------- PYTHON 3 --------
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # O(n) time | O(n) space

        # Hashmap to store numbers and their indexes
        # { number: index }
        storage = {} # O(n) space
        
        for i in range(len(nums)): # O(n) time
            # complement -> Number needed to add to current 
            # element to reach the target
            complement = target - nums[i]
            
            # If the complement is in the storage we can add up to
            # target -> RETURN indexes
            if complement in storage:
                return [storage[complement], i]
            
            # Store current element with his index
            storage[nums[i]] = i       